python/mspasspy/db/database.py

```diff
@@ -574,45 +645,182 @@ def save_data(self, mspass_object, mode="promiscuous", storage_mode='gridfs', di
+                if 'data_tag' in insertion_dict:
+                    insertion_dict.erase('data_tag')
+            # finally ready to insert the wf doc - keep the id as we'll need
+            # it for tagging any elog entries
+            wfid = wf_collection.insert_one(insertion_dict).inserted_id
+            # Put wfid into the object's meta as the new definition of 
+            # the parent of this waveform
+            # TODO - I think we need to reset the history container here 
+            # and mark this new save as an ORIGIN leaving till later
+            mspass_object['_id']=wfid
+
+            # Empty error logs are skipped.  When nonzero tag them with tid
+            # just returned
+            if mspass_object.elog.size() > 0:
+                elog_id_name = self.database_schema.default_name('elog') + '_id'
+                # _save_elog uses a  null id as a signal to add a new record
+                # When we land here the record must be new since it is
+                # associated with a new wf document.  elog_id=None is default
+                # but set it explicitly for clarity
+                elog_id = self._save_elog(mspass_object, elog_id=None)
+            else:
+                elog_id = None
+            # When history is enable we need to do an update to put the
+            # wf collection id as a cross-reference.    Any value stored
+            # above with saave_history may be incorrect.  We use a
+            # stock test with the is_empty method for know if history data is present
+            if not mspass_object.is_empty():
+                history_object_col = self[self.database_schema.default_name('history_object')]
+                wf_id_name = wf_collection_name + '_id'
+                filter_ = {'_id': history_object_id}
+                update_dict={wf_id_name : wfid}
+                history_object_col.update_one(filter_, {'$set': update_dict})
+            return [wfid,history_object_id,elog_id]
 
         else:
-            # FIXME: we could have recorded the full stack here, but need to revise the logger object
-            # to make it more powerful for Python logging.
-            mspass_object.elog.log_verbose(
-                sys._getframe().f_code.co_name, "Skipped saving dead object")
-            self._save_elog(mspass_object)
+            # We land here when the input is dead or was killed during a
+            # cautious or pedantic mode edit of the metadata.
+            elog_id_name = self.database_schema.default_name('elog') + '_id'
+            if elog_id_name in mspass_object:
+                old_elog_id=mspass_object[elog_id_name]
+            else:
+                old_elog_id=None
+            elog_id = self._save_elog(mspass_object,elog_id=old_elog_id)
+            if '_id' in mspass_object:
+                dead_wf_id=mspass_object['_id']
+            else:
+                dead_wf_id=None
+            return [dead_wf_id,None,elog_id]
 
```

Comment:
I think save_data is much cleaner than before and it would work well when we save a mspass_object for the first time. However,  I could think of a scenario that a user may call save_data for a mspass_object for many times. In that case, some of the logic in the code is not applicable because we assume when the user call save_data, it's always a new wf document. Also, there would be a lot of wf documents associated with this mspass_object in the MongoDB database. I am not sure if these would cause some problems during our processing workflow and I think I would let you guys for discussions.

Reply:
Definitely worth a discussion. To complex to answer coherently here at this time.